### PR TITLE
Fix isRoundingError

### PR DESCRIPTION
**Pull request by abandeali1**
- Closes issues #98 and #92

### CODE REVIEW

File: test/ts/exchange/helpers.ts [hunk 1]
```diff
@@ -76,4 +77,56 @@ contract('Exchange', (accounts: string[]) => {
...
+    it('should round down', async () => {
+      const numerator = new BigNumber(2);
+      const denominator = new BigNumber(3);
+      const target = new BigNumber(10);
+
+      const partialAmount = await exWrapper.getPartialAmountAsync(numerator, denominator, target);
+      const expectedPartialAmount = '6';
+      assert.equal(partialAmount.toString(), expectedPartialAmount);
+    });
+  });
```
Comment (LogvinovLeon) [inline test/ts/exchange/helpers.ts]: I'll add one more test for rounding of 0.5. Different implementations differ and we want to make sure it's behaving like expected.
```
const numerator = new BigNumber(1);
const denominator = new BigNumber(20);
const target = new BigNumber(10);
```

File: test/ts/exchange/helpers.ts [hunk 2]
```diff
@@ -76,4 +77,56 @@ contract('Exchange', (accounts: string[]) => {
...
+      assert.equal(isRoundingError, false);
+    });
+
+    it('should return false when there is rounding error <= 0.1%', async () => {
+      // randomly generated numbers
+      const numerator = new BigNumber(76564);
+      const denominator = new BigNumber(676373677);
+      const target = new BigNumber(105762562);
+      // rounding error = ((76564*105762562/676373677) - floor(76564*105762562/676373677)) / (76564*105762562/676373677) = 0.0007%
+      const isRoundingError = await exWrapper.isRoundingErrorAsync(numerator, denominator, target);
```
Comment (LogvinovLeon) [inline test/ts/exchange/helpers.ts]: I'll rename `exWrapper` to `exchangeWrapper`

File: contracts/Exchange.sol [hunk 3]
```diff
@@ -490,7 +490,20 @@ contract Exchange is SafeMath {
...
+                safeMul(numerator, target),
+                1000000
+            ),
+            denominator
+        );
+        uint errPercentageTimes1000 = safeDiv(
+            safeSub(partialAmountWithoutErr, partialAmountWithErr), // Absolute rounding error, times 1,000,000
+            safeDiv(partialAmountWithoutErr, 1000)                  // Amount being filled, times 1,000
+        );
+        return errPercentageTimes1000 > 1;
```
Comment (LogvinovLeon) [inline contracts/Exchange.sol]: In contrast with the mutisig's bytes to bytes4 this function is executed on every fill and that's why gas cost matters. What's the gas cost of this one?
Can we somehow optimize it?
Maybe it's better to leave it for v2.

File: contracts/Exchange.sol [hunk 4]
```diff
@@ -490,7 +490,20 @@ contract Exchange is SafeMath {
...
+            safeMul(
+                safeMul(numerator, target),
+                1000000
+            ),
+            denominator
+        );
+        uint errPercentageTimes1000 = safeDiv(
+            safeSub(partialAmountWithoutErr, partialAmountWithErr), // Absolute rounding error, times 1,000,000
+            safeDiv(partialAmountWithoutErr, 1000)                  // Amount being filled, times 1,000
+        );
```
Comment (LogvinovLeon) [inline contracts/Exchange.sol]: With all the safeDivs and safeSubs this logic is cryptic.
I'll add a comment explaining what's happening similar to the one we have in tests.

File: contracts/Exchange.sol (hunk 3 above)
Comment (abandeali1) [inline contracts/Exchange.sol]: I ran a few tests and the difference seems to be negligible, surprisingly. Probably worth looking into a bit more.

File: test/ts/exchange/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+      const isRoundingError = await exchangeWrapper.isRoundingErrorAsync(numerator, denominator, target);
+      assert.equal(isRoundingError, false);
+    });
+
+    it('should return false when there is rounding error <= 0.1%', async () => {
+      // randomly generated numbers
+      const numerator = new BigNumber(76564);
+      const denominator = new BigNumber(676373677);
+      const target = new BigNumber(105762562);
+      // rounding error = ((76564*105762562/676373677) - floor(76564*105762562/676373677)) / (76564*105762562/676373677) = 0.0007%
```
Comment (ethers) [inline test/ts/exchange/helpers.ts]: is this computation supposed to be the same as https://github.com/0xProject/contracts/issues/98 ?

**Pull request merged**